File: multiple_dispatch_vs_oop/md.jl
abstract type Boundary end

struct BoundaryDefault <: Boundary
    a::Array{Float64, 2}
    b::Array{Float64, 2}

    BoundaryDefault(itot, jtot) = new(zeros(itot, jtot), zeros(itot, jtot))
end

struct BoundarySurface <: Boundary
    a::Array{Float64, 2}
    b::Array{Float64, 2}
    c::Array{Float64, 2}

    BoundarySurface(itot, jtot) = new(zeros(itot, jtot), zeros(itot, jtot), zeros(itot, jtot))
end

function process_surface(b::BoundaryDefault)
end

function process_surface(b::BoundarySurface)
    b.c[:, :] .+= 100
end

function process_boundaries(b::Boundary)
    b.a[:, :] .+= 1
    b.b[:, :] .+= 10

    process_surface(b)
end

b_default = BoundaryDefault(2, 2)
b_surface = BoundarySurface(2, 2)

process_boundaries(b_default)
process_boundaries(b_surface)

println("b_default: $b_default")
println("b_surface: $b_surface")

File: multiple_dispatch_vs_oop/md.cpp
#include <vector>
#include <algorithm>

struct Boundary
{
    virtual ~Boundary() {};
    std::vector<double> a;
    std::vector<double> b;
};

struct BoundaryDefault : public Boundary
{
};

struct BoundarySurface : public Boundary
{
    std::vector<double> c;
};

void process_surface(BoundaryDefault& b)
{
}

void process_surface(BoundarySurface& b)
{
    for (double& d : b.c)
        d += 100.;
}

void process_boundaries(Boundary& b)
{
    for (double& d : b.a)
        d += 1.;
    for (double& d : b.b)
        d += 10.;

    if (auto b_cast = dynamic_cast<BoundaryDefault*>(&b))
        process_surface(*b_cast);
    else if (auto b_cast = dynamic_cast<BoundarySurface*>(&b))
        process_surface(*b_cast);
}

int  main()
{
    BoundaryDefault b_default;
    BoundarySurface b_surface;

    process_boundaries(b_default);
    process_boundaries(b_surface);

    return 0;
}
